Comprehensive type checking for `from_pretrained` kwargs

guiyrt opened the pull request:
# What does this PR do?
### Changes
- Moved type checking to just before pipeline instantiation, so all kwargs are checked
- Full type-check for collections (list, dicts, ...), every element checked
- More detailed warning for unexpected arguments type (`List[ControlNetModel]` instead of `list`)

### To-do
- [X] Where should the new functions `is_valid_type` and `get_detailed_type` be placed?
- [ ] ~~According to new functions location, add simple tests for type checking.~~

These changes are proposed based on testing for #10747.

Example warning when providing `controlnet`as `List[ControlNetUnionModel]` for `StableDiffusionXLControlNetPipeline`, where `List[ControlNetModel]` is expected:

```
Expected types for controlnet: (<class 'diffusers.models.controlnets.controlnet.ControlNetModel'>,
typing.List[diffusers.models.controlnets.controlnet.ControlNetModel], 
typing.Tuple[diffusers.models.controlnets.controlnet.ControlNetModel],
<class 'diffusers.models.controlnets.multicontrolnet.MultiControlNetModel'>),
got typing.List[diffusers.models.controlnets.controlnet_union.ControlNetUnionModel].
```

<details>
<summary>Code for warning replication</summary>

```python
import torch

from diffusers import StableDiffusionXLControlNetPipeline
from diffusers.models import ControlNetUnionModel, AutoencoderKL
from diffusers.utils import load_image


controlnet = ControlNetUnionModel.from_pretrained(
    "brad-twinkl/controlnet-union-sdxl-1.0-promax", torch_dtype=torch.float16
)

vae = AutoencoderKL.from_pretrained(
    "madebyollin/sdxl-vae-fp16-fix", torch_dtype=torch.float16
)

pipe = StableDiffusionXLControlNetPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0",
    controlnet=[controlnet, controlnet],
    vae=vae,
    torch_dtype=torch.float16,
    variant="fp16",
)

room_seg_img = load_image("https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/sd_controlnet/room_seg.png")
pose_img = load_image("https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/sd_controlnet/pose.png")


pipe.enable_model_cpu_offload()

image = pipe(
    prompt="an astronaut in a space station",
    width=1024,
    height=1024,
    negative_prompt="lowres, low quality, worst quality",
    generator=torch.manual_seed(42),
    guidance_scale=5,
    num_inference_steps=50,
    image=[pose_img, room_seg_img],
).images[0]

image.save("result.jpg")
```
</details>

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->



## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [X] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [X] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@hlky 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @.

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Core library:

- Schedulers: @yiyixuxu
- Pipelines and pipeline callbacks: @yiyixuxu and @asomoza
- Training examples: @sayakpaul
- Docs: @stevhliu and @sayakpaul
- JAX and MPS: @pcuenca
- Audio: @sanchit-gandhi
- General functionalities: @sayakpaul @yiyixuxu @DN6

Integrations:

- deepspeed: HF Trainer/Accelerate: @SunMarc
- PEFT: @sayakpaul @BenjaminBossan

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- transformers: [different repo](https://github.com/huggingface/transformers)
- safetensors: [different repo](https://github.com/huggingface/safetensors)

-->

hlky reviewed (approved):
Thanks @guiyrt, nice work. Could you take a look through [pipeline test output](https://github.com/huggingface/diffusers/actions/runs/13243381033/job/36965112365?pr=10758) searching for `Expected types for ` (GitHub's built in search works best)

There are some easy cases that we could fix like

```
Expected types for feature_extractor: (<class 'transformers.models.clip.image_processing_clip.CLIPImageProcessor'>,), got <class 'transformers.models.clip.feature_extraction_clip.CLIPFeatureExtractor'>.
```

For that we could do global find+replace on `feature_extractor: CLIPImageProcessor` -> `feature_extractor: CLIPFeatureExtractor`.

and some that need investigating

```
Expected types for unet: (<class 'inspect._empty'>,), got <class 'diffusers.models.unets.unet_2d.UNet2DModel'>.
```

Type correctness is not strictly enforced so some warnings are expected but we should make a best effort to reduce the number of new warnings that we're introducing. If we find a particular component to be a problem we can skip it like `scheduler`.

hlky commented:
Failing tests appear unrelated, will re-run later.

guiyrt commented:
@hlky Findings from looking through the test logs

### 1. Using XYZFast `tokenizer` when only XYZ is annotated (and vice-versa)

We can make a quick search and replace and update all tokenizer annotations to be `Union[XYZBase, XYZFast]`, but as this is a big change, let me know if you agree. 

18 occurrences
```
Expected types for tokenizer: (<class 'transformers.models.xlm_roberta.tokenization_xlm_roberta.XLMRobertaTokenizer'>,), 
got <class 'transformers.models.xlm_roberta.tokenization_xlm_roberta_fast.XLMRobertaTokenizerFast'>.
```

68 occurrences
```
Expected types for tokenizer: (<class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>,),
got <class 'transformers.models.t5.tokenization_t5_fast.T5TokenizerFast'>.
```

9 occurrences
```
Expected types for tokenizer: (<class 'transformers.models.bert.tokenization_bert.BertTokenizer'>,),
got <class 'transformers.models.bert.tokenization_bert_fast.BertTokenizerFast'>.
```

4 occurences
```
Expected types for tokenizer: (<class 'transformers.models.llama.tokenization_llama_fast.LlamaTokenizerFast'>,), got <class 'transformers.models.llama.tokenization_llama.LlamaTokenizer'>.
```

### 2. `CLIPFeatureExtractor` as `feature_extractor`

This comes from tests that use a hf-internal-testing repo with legacy `CLIPFeatureExtractor` instead of `CLIPImageProcessor`. A warning from `transformers` is also thrown ```FutureWarning: The class CLIPFeatureExtractor is deprecated and will be removed in version 5 of Transformers. Please use CLIPImageProcessor instead.```

15 occurrences
```
Expected types for feature_extractor: (<class 'transformers.models.clip.image_processing_clip.CLIPImageProcessor'>,),
got <class 'transformers.models.clip.feature_extraction_clip.CLIPFeatureExtractor'>.
```

https://github.com/huggingface/diffusers/blob/8ae8008b0d096d2b093f5b7c660715a93f74f17a/tests/pipelines/test_pipelines.py#L582-L583

### 3. `PipelineFastTests::test_optional_components`

This test purposefully sets `requires_safety_checker` as `[True, True]` and `safety_checker` as an unet and `feature_extractor` as a function, so the warnings here are expected. 

2+2+2 occurrences
```
Expected types for safety_checker: (<class 'diffusers.pipelines.stable_diffusion.safety_checker.StableDiffusionSafetyChecker'>,),
got <class 'diffusers.models.unets.unet_2d_condition.UNet2DConditionModel'>.
Expected types for requires_safety_checker: (<class 'bool'>,), got typing.List[bool].
Expected types for feature_extractor: (<class 'transformers.models.clip.image_processing_clip.CLIPImageProcessor'>,),
got <class 'function'>.
```

https://github.com/huggingface/diffusers/blob/8ae8008b0d096d2b093f5b7c660715a93f74f17a/tests/pipelines/test_pipelines.py#L1695-L1701

### 4. Missing type hinting

Added the intended types.

6+6 occurrences from `HunyuanDiTPipeline`s.
```
Expected types for text_encoder_2: (<class 'inspect._empty'>,),
got <class 'transformers.models.t5.modeling_t5.T5EncoderModel'>.
Expected types for tokenizer_2: (<class 'inspect._empty'>,),
got <class 'transformers.models.t5.tokenization_t5_fast.T5TokenizerFast'>.
```

https://github.com/huggingface/diffusers/blob/8ae8008b0d096d2b093f5b7c660715a93f74f17a/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py#L210-L211

https://github.com/huggingface/diffusers/blob/57ac6738028004143cf19c362a81d7d135d1de24/src/diffusers/pipelines/controlnet_hunyuandit/pipeline_hunyuandit_controlnet.py#L235-L236

7+7 occurrences from `CustomPipeline` tests. Only showed for `unet` because `scheduler` is not checked.
```
Expected types for unet: (<class 'inspect._empty'>,),
got <class 'diffusers.models.unets.unet_2d.UNet2DModel'>.
Expected types for unet: (<class 'inspect._empty'>,),
got <class 'diffusers.models.unets.unet_1d.UNet1DModel'>.
```

https://github.com/huggingface/diffusers/blob/8ae8008b0d096d2b093f5b7c660715a93f74f17a/tests/fixtures/custom_pipeline/pipeline.py#L36

https://github.com/huggingface/diffusers/blob/57ac6738028004143cf19c362a81d7d135d1de24/src/diffusers/pipelines/ddim/pipeline_ddim.py#L50

https://github.com/huggingface/diffusers/blob/57ac6738028004143cf19c362a81d7d135d1de24/src/diffusers/pipelines/ddpm/pipeline_ddpm.py#L50

https://github.com/huggingface/diffusers/blob/57ac6738028004143cf19c362a81d7d135d1de24/src/diffusers/pipelines/deprecated/repaint/pipeline_repaint.py#L94

https://github.com/huggingface/diffusers/blob/57ac6738028004143cf19c362a81d7d135d1de24/src/diffusers/pipelines/dance_diffusion/pipeline_dance_diffusion.py#L52


### 5. `CustomPipelineTests`

Not sure what to make of this.

2+2+2 occurrences
```
Expected types for unet: (<class 'diffusers_modules.local.unet.my_unet_model.MyUNetModel'>,),
got <class 'diffusers_modules.local.my_unet_model.MyUNetModel'>.
Expected types for unet: (<class 'diffusers.models.unets.unet_2d_condition.UNet2DConditionModel'>,),
got <class 'diffusers_modules.local.my_unet_model.MyUNetModel'>.
Expected types for unet: (<class 'diffusers_modules.local.unet.my_unet_model.MyUNetModel'>,),
got <class 'diffusers_modules.local.my_unet_model.MyUNetModel'>.
```

### 6. Incomplete `unet` type hints in `AnimateDiffVideoToVideoPipeline`s

Changed to be `unet: Union[UNet2DConditionModel, UNetMotionModel]`, as in `AnimateDiffSDXLPipeline`.

8 occurrences 
```
Expected types for unet: (<class 'diffusers.models.unets.unet_2d_condition.UNet2DConditionModel'>,),
got <class 'diffusers.models.unets.unet_motion_model.UNetMotionModel'>.
```

https://github.com/huggingface/diffusers/blob/8ae8008b0d096d2b093f5b7c660715a93f74f17a/src/diffusers/pipelines/animatediff/pipeline_animatediff_video2video_controlnet.py#L249

https://github.com/huggingface/diffusers/blob/8ae8008b0d096d2b093f5b7c660715a93f74f17a/src/diffusers/pipelines/animatediff/pipeline_animatediff_video2video.py#L227


### 7. Subclasses not checked in `is_valid_type`

This should be correct, so changed `is_valid_type` to use `isinstance` to also allow subsclasses. This way, you can annotate with a parent class.

4 occurrences
```
Expected types for bert: (<class 'transformers.modeling_utils.PreTrainedModel'>,),
got <class 'transformers.models.clip.modeling_clip.CLIPTextModel'>.
```


### 8. Type hinting with `AutoTokenizer` and `AutoModel` on `SanaPipelines`s

The proper base classes are `PreTrainedModel` and `PreTrainedTokenizerBase`.

7 occurrences
```
Expected types for tokenizer: (<class 'transformers.models.auto.tokenization_auto.AutoTokenizer'>,),
got <class 'transformers.models.gemma.tokenization_gemma.GemmaTokenizer'>.
```

3 occurences
```
Expected types for tokenizer: (<class 'transformers.models.auto.tokenization_auto.AutoTokenizer'>,),
got <class 'transformers.models.gemma.tokenization_gemma_fast.GemmaTokenizerFast'>.
```

3 occurrences
```
Expected types for text_encoder: (<class 'transformers.models.auto.modeling_auto.AutoModel'>,),
got <class 'transformers.models.gemma.modeling_gemma.GemmaForCausalLM'>.
```

3 occurrences
```
Expected types for text_encoder: (<class 'transformers.models.auto.modeling_auto.AutoModelForCausalLM'>,),
got <class 'transformers.models.gemma2.modeling_gemma2.Gemma2ForCausalLM'>.
```

4 occurrences
```
Expected types for text_encoder: (<class 'transformers.models.auto.modeling_auto.AutoModelForCausalLM'>,),
got <class 'transformers.models.gemma2.modeling_gemma2.Gemma2Model'>.
```

https://github.com/huggingface/diffusers/blob/57ac6738028004143cf19c362a81d7d135d1de24/src/diffusers/pipelines/sana/pipeline_sana.py#L203-L204

https://github.com/huggingface/diffusers/blob/57ac6738028004143cf19c362a81d7d135d1de24/src/diffusers/pipelines/pag/pipeline_pag_sana.py#L163-L164


### 9. Interchanged use of `CLIPTextModel` and `CLIPTextModelWithProjection`

Just swapped with the intended type.

4 occurrences
```
Expected types for text_encoder: (<class 'transformers.models.clip.modeling_clip.CLIPTextModelWithProjection'>,),
got <class 'transformers.models.clip.modeling_clip.CLIPTextModel'>.
```

https://github.com/huggingface/diffusers/blob/57ac6738028004143cf19c362a81d7d135d1de24/src/diffusers/pipelines/stable_diffusion/pipeline_stable_unclip.py#L144

8 occurrences
```
Expected types for text_encoder: (<class 'transformers.models.clip.modeling_clip.CLIPTextModel'>,),
got <class 'transformers.models.clip.modeling_clip.CLIPTextModelWithProjection'>.
```

https://github.com/huggingface/diffusers/blob/8ae8008b0d096d2b093f5b7c660715a93f74f17a/src/diffusers/pipelines/stable_cascade/pipeline_stable_cascade.py#L96

https://github.com/huggingface/diffusers/blob/8ae8008b0d096d2b093f5b7c660715a93f74f17a/src/diffusers/pipelines/stable_cascade/pipeline_stable_cascade_combined.py#L79

4 occurrences
```
Expected types for prior_text_encoder: (<class 'transformers.models.clip.modeling_clip.CLIPTextModel'>,),
got <class 'transformers.models.clip.modeling_clip.CLIPTextModelWithProjection'>.
```

https://github.com/huggingface/diffusers/blob/8ae8008b0d096d2b093f5b7c660715a93f74f17a/src/diffusers/pipelines/stable_cascade/pipeline_stable_cascade_combined.py#L84

guiyrt commented:
Found another warning related to custom pipelines, this time on "hf-internal-testing/diffusers-dummy-pipeline". The fix is having the correct type hinting there.

4 occurrences
```
Expected types for unet: (<class 'inspect._empty'>,),
got <class 'diffusers.models.unets.unet_2d.UNet2DModel'>
```

https://github.com/huggingface/diffusers/blob/067eab1b3aaf4d09f85edf21d8b147e0980c662a/tests/pipelines/test_pipelines.py#L1043-L1053

guiyrt commented:
I opened PRs on the hf-internal-testing repos with warnings:
[Replacing deprecated `CLIPFeatureExtractor` for `CLIPImageProcessor` 1](https://huggingface.co/hf-internal-testing/tiny-stable-diffusion-pipe-variants-right-format/discussions/1)
[Replacing deprecated `CLIPFeatureExtractor` for `CLIPImageProcessor` 2](https://huggingface.co/hf-internal-testing/tiny-stable-diffusion-pipe-variants-right-format/discussions/2)
[Added type annotations for pipeline init args](https://huggingface.co/hf-internal-testing/diffusers-dummy-pipeline/discussions/3)

The last one is regarding arguments with no annotations, which might be common on custom pipelines? To keep warnings relevant, it might be a good idea to skip type checking if there is no type annotated for a given argument.

hlky commented:
1. Maybe we just don't check tokenizer
2. If it's an internal testing checkpoint the warning is not important
3. If it's expected it's ok
4. Thanks
5. Custom pipeline so not important
6. Thanks
7. Annotations of parent class `PreTrainedModel` are not that useful, changing to correct type be left as a todo
8. As above
9. Thanks

guiyrt commented:
> 1. Maybe we just don't check tokenizer

`tokenizer` is now skipped

> 7. Annotations of parent class `PreTrainedModel` are not that useful, changing to correct type be left as a todo

This to-do includes:
- [ ] `LDMTextToImagePipeline`:  Supposedly the types in the docs are `LDMBertModel` and `BertTokenizer`, but text_encoder and tokenizer used in the tests are `CLIPTextModel` and `CLIPTokenizer`.
- [ ] `LuminaText2ImgPipeline` and `Lumina2Text2ImgPipeline`:  Docs mention T5, but tests use Gemma.
- [X] `SanaPAGPipeline`and `SanaPipeline`: From what I understood and the tests, these use GemmaTokenizer[Fast] and `Gemma2Model` for `SanaPipeline` tests and `Gemma2CausalLM` for `SanaPAGPipeline`, so I annotated text_encoder with `Gemma2PreTrainedModel` type, as it works for both.

guiyrt commented:
@hlky can we move the functions `get_detailed_type` and `is_valid_type` to some utils file? Maybe a new typing_utils.py or something? I don't think they should stay in the middle of `DiffusionPipeline::from_pretrained`

hlky commented:
`LDMTextToImagePipeline` I'm not sure, could be that it supports both types or incorrect type hint.

`LuminaText2ImgPipeline` Yeah Lumina is Gemma, docstring/type hint would have been copied from some other pipeline.

`SanaPAGPipeline` and `SanaPipeline` is Gemma, not sure if it should be `Gemma2Model` or `Gemma2ForCausalLM` though and I'm assuming it's supposed to be the same for both. Using `Gemma2PreTrainedModel` should be ok.

@yiyixuxu WDYT about `typing_utils.py`? Might be some other code that could be moved there, these functions probably won't be used elsewhere though. IMO I don't mind the functions being in `from_pretrained`, if I'm working on `from_pretrained` I've got all the context of those functions immediately available.

guiyrt commented:
> `LuminaText2ImgPipeline` Yeah Lumina is Gemma, docstring/type hint would have been copied from some other pipeline.

I updated the docs. But if I got it right, Lumina v1 uses `GemmaModel` and v2 uses `Gemma2Model`, however the FastTests of both used `GemmaForCausalLM`. For Lumina v1 we can annotate as `GemmaPreTrainedModel`, but if we annotated Lumina v2 with `Gemma2PreTrainedModel` it would produce a warning for the tests. So on top of updating the type annotations, I also updated the tests for Lumina v2, it was easy because there is no comparison with expected output hardcoded. Ran it locally and passed :)

hlky commented:
Just temporary issues, happens sometimes. Thanks for all the iterations on this @guiyrt, should be good to go after @yiyixuxu's comments on whether to add `typing_utils.py`.

hlky commented:
Gentle ping @yiyixuxu

yiyixuxu commented:
thanks for the PR @guiyrt @hlky 
it can to into this util file I think! https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pipeline_loading_utils.py

hlky commented:
Failing tests are unrelated.

Thanks @guiyrt
